server/src/main/java/org/elasticsearch/script/field/ValidatingMap.java

```diff
@@ -0,0 +1,259 @@
+            return number;
+        }
+        throw new IllegalArgumentException(
+            "unexpected type for [" + key + "] with value [" + value + "], expected Number, got [" + value.getClass().getName() + "]"
+        );
+    }
+
+    /**
+     * Get the validator for the given key, or the identity function if the key has no validator.
+     */
+    protected BiFunction<String, Object, Object> getValidator(String key) {
+        return validators.getOrDefault(key, IDENTITY);
+    }
+
+    /**
+     * Set of entries of the wrapped map that calls the appropriate validator before changing an entries value or removing an entry.
+     *
+     * Inherits {@link AbstractSet#removeAll(Collection)}, which calls the overridden {@link #remove(Object)} which performs validation.
+     *
+     * Inherits {@link AbstractCollection#retainAll(Collection)} and {@link AbstractCollection#clear()}, which both use
+     * {@link EntrySetIterator#remove()} for removal.
+     */
+    class EntrySet extends AbstractSet<Map.Entry<String, Object>> {
+        Set<Map.Entry<String, Object>> set;
+
+        EntrySet(Set<Map.Entry<String, Object>> set) {
+            this.set = set;
+        }
+
+        @Override
+        public Iterator<Map.Entry<String, Object>> iterator() {
+            return new EntrySetIterator(set.iterator());
+        }
+
+        @Override
+        public int size() {
+            return set.size();
+        }
+
+        @Override
+        public boolean remove(Object o) {
+            if (o instanceof Map.Entry<?, ?> entry) {
+                if (entry.getKey()instanceof String key) {
+                    getValidator(key).apply(key, null);
+                }
+                return set.remove(o);
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Iterator over the wrapped map that returns a validating {@link Entry} on {@link #next()} and validates on {@link #remove()}.
+     *
+     * {@link #remove()} is called by remove in {@link AbstractMap#values()}, {@link AbstractMap#keySet()}, {@link AbstractMap#clear()} via
+     * {@link AbstractSet#clear()}
+     */
+    class EntrySetIterator implements Iterator<Map.Entry<String, Object>> {
+        Iterator<Map.Entry<String, Object>> iterator;
+        Entry cur;
+
+        EntrySetIterator(Iterator<Map.Entry<String, Object>> iterator) {
+            this.iterator = iterator;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return iterator.hasNext();
+        }
+
+        @Override
+        public Map.Entry<String, Object> next() {
+            return cur = new Entry(iterator.next());
+        }
+
+        /**
+         * Remove current entry from the backing Map.  Checks the Entry's key's validator, if one exists, before removal.
+         * @throws IllegalArgumentException if the validator does not allow the Entry to be removed
+         */
+        @Override
+        public void remove() {
```

Comment:
Wouldn't the iterator remove eventually delegate to a remove on the map, which will run the validator? Or why is a validator run at all? It seems like a hacky way to disallow removal.

Reply:
There is this map and the wrapped map.  This iterator is needed because `AbstractSet` requires subclasses to implement `iterator()`.  If we just take the underlying maps entry set, the validators will not be called.

Reply:
Ok, but why are the validators run on removal?

Reply:
You cannot remove `_version`.  If scripts remove `_version` then either we need to return a sentinel value from `getVersion()` (`Long.MIN_VALUE`  or similar) or we have to throw an `IllegalArgumentException`

Reply:
Validators are still run on removal because a validator may raise on removal, however this change does not have any such validator.